### PR TITLE
[Transform]Move ReservedFnFunctionRector to Php74 namespace

### PR DESCRIPTION
**Pull request by samsonasik**
With direct update `function fn()` to `function f()`, if `function f()` exists, increment the name.

per https://github.com/rectorphp/rector-src/pull/1423#issuecomment-989899266

### CODE REVIEW

**Comment by samsonasik** [comment]
@TomasVotruba when running test read `function fn()`, it got error:

```bash
There was 1 error:

1) Rector\Tests\Php74\Rector\Function_\ReservedFnFunctionRector\ReservedFnFunctionRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\Parser\ParserErrorsException: Syntax error, unexpected T_FN, expecting T_STRING or '(' on line 9, Syntax error, unexpected '{', expecting T_DOUBLE_ARROW on line 10, Syntax error, unexpected T_LNUMBER, expecting T_VARIABLE on line 14

phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Parser/RichParser.php:42
/Users/samsonasik/www/rector-src/src/PhpParser/Parser/RectorParser.php:26
/Users/samsonasik/www/rector-src/src/PhpParser/Parser/RectorParser.php:31
/Users/samsonasik/www/rector-src/src/Application/FileProcessor.php:28
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:128
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:52
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:73
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:42
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:177
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:131
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/Users/samsonasik/www/rector-src/rules-tests/Php74/Rector/Function_/ReservedFnFunctionRector/ReservedFnFunctionRectorTest.php:18
```

is `RichParser` may need to be re-configured to support old php 7.x syntax? it may related with issue: 

- https://github.com/rectorphp/rector/issues/6799 

which `Match` is only reserved keyword in php 8.x

**Comment by samsonasik** [comment]
@TomasVotruba I rebased, it seems tests still got error:

```bash
There was 1 error:

1) Rector\Tests\Php74\Rector\Function_\ReservedFnFunctionRector\ReservedFnFunctionRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\Parser\ParserErrorsException: Syntax error, unexpected T_FN, expecting T_STRING or '(' on line 9, Syntax error, unexpected '{', expecting T_DOUBLE_ARROW on line 10, Syntax error, unexpected T_LNUMBER, expecting T_VARIABLE on line 14

phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Parser/RichParser.php:42
/Users/samsonasik/www/rector-src/src/PhpParser/Parser/RectorParser.php:26
/Users/samsonasik/www/rector-src/src/PhpParser/Parser/RectorParser.php:31
/Users/samsonasik/www/rector-src/src/Application/FileProcessor.php:28
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:128
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:52
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:73
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:42
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:177
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:131
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/Users/samsonasik/www/rector-src/rules-tests/Php74/Rector/Function_/ReservedFnFunctionRector/ReservedFnFunctionRectorTest.php:18
```

**Comment by villfa** [comment]
I was writing a new Rector rule and got the exact same error:
```
1) Rector\Tests\Php80\Rector\Namespace_\NamespaceAsSingleTokenRector\NamespaceAsSingleTokenRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\Parser\ParserErrorsException: Syntax error, unexpected T_NS_SEPARATOR, expecting ';' or '{' on line 3

phar:///home/runner/work/rector-src/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Parser/RichParser.php:42
/home/runner/work/rector-src/rector-src/src/PhpParser/Parser/RectorParser.php:26
/home/runner/work/rector-src/rector-src/src/PhpParser/Parser/RectorParser.php:31
/home/runner/work/rector-src/rector-src/src/Application/FileProcessor.php:28
/home/runner/work/rector-src/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:128
/home/runner/work/rector-src/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:52
/home/runner/work/rector-src/rector-src/src/Application/ApplicationFileProcessor.php:73
/home/runner/work/rector-src/rector-src/src/Application/ApplicationFileProcessor.php:42
/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:177
/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:131
/home/runner/work/rector-src/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/home/runner/work/rector-src/rector-src/rules-tests/Php80/Rector/Namespace_/NamespaceAsSingleTokenRector/NamespaceAsSingleTokenRectorTest.php:18
```

The code I try to transform is valid only with PHP 7, but RichParser is instantiated with the current PHP version, PHP 8, hence the syntax error.

Is there a way to configure RichParser from `config/configured_rule.php`?

**Comment by villfa** [comment]
> Is there a way to configure RichParser from `config/configured_rule.php`?

I think I've found an example here: https://github.com/rectorphp/rector-src/blob/c7638288f7461f46c57307a22b9454c46cfa3cbc/build/config/config-downgrade.php#L25

**Comment by samsonasik** [comment]
Linked to https://github.com/rectorphp/rector-src/pull/1503 for possible usage of php 7 parser.

File: src/PhpParser/Parser/RectorParser.php [hunk 1]
```diff
@@ -23,7 +35,22 @@ public function __construct(
...
+            $message = $parserErrorsException->getMessage();
+
+            if (! StringUtils::isMatch($message, self::FN_ERROR_REGEX)) {
+                throw $parserErrorsException;
+            }
+
+            $parser = $this->parser;
+
+            $lexer = &\Closure::bind(static fn&($parser) => $parser->lexer, null, $parser)($parser);
+            $lexer = new Emulative(['phpVersion' => Emulative::PHP_7_3]);
```
Comment (samsonasik) [inline src/PhpParser/Parser/RectorParser.php]: @TomasVotruba it possibly needs to update `Emulative` Lexer instead, I tried this, but got error:

```bash
There was 1 error:

1) Rector\Tests\Php74\Rector\Function_\ReservedFnFunctionRector\ReservedFnFunctionRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
TypeError: PhpParser\Lexer::getTokens(): Return value must be of type array, null returned
```

Comment (samsonasik) [reply src/PhpParser/Parser/RectorParser.php]: When I update the code to be like this:

```php

            $parser = new Php7(new Emulative([
                'phpVersion' => Emulative::PHP_7_3,
            ]));

            return $parser->parse(file_get_contents($smartFileInfo->getRealPath()));
```

It got error:

```bash
There was 1 error:

1) Rector\Tests\Php74\Rector\Function_\ReservedFnFunctionRector\ReservedFnFunctionRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\Broker\ClassNotFoundException: Class Fixture was not found while trying to analyse it - discovering symbols is probably not configured properly.

phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ChainReflectionProvider.php:40
phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/MemoizingReflectionProvider.php:39
phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:485
phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:253
phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:468
phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php:221
/Users/samsonasik/www/rector-src/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:122
/Users/samsonasik/www/rector-src/packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php:109
/Users/samsonasik/www/rector-src/packages/NodeTypeResolver/NodeScopeAndMetadataDecorator.php:36
/Users/samsonasik/www/rector-src/src/Application/FileProcessor.php:34
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:128
/Users/samsonasik/www/rector-src/src/Application/FileProcessor/PhpFileProcessor.php:52
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:73
/Users/samsonasik/www/rector-src/src/Application/ApplicationFileProcessor.php:42
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:177
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:131
/Users/samsonasik/www/rector-src/packages/Testing/PHPUnit/AbstractRectorTestCase.php:115
/Users/samsonasik/www/rector-src/rules-tests/Php74/Rector/Function_/ReservedFnFunctionRector/ReservedFnFunctionRectorTest.php:18
```

**Comment by TomasVotruba** [comment]
What is missing here to be finished?

**Comment by samsonasik** [comment]
It still fail to get php-parser work on php 7.x to verify `fn` as valid function, using emulative still error, ref https://github.com/rectorphp/rector-src/pull/1471#discussion_r769679182

**Comment by TomasVotruba** [comment]
I see. Could you separate the rule moving and the parser update? E.g. by skipping test.
The first one is simple and will clear the rector reports about configure(). 
Tnd the latter will take more time to resolve.

**Comment by samsonasik** [comment]
Ok, I reverted RectorParser change https://github.com/rectorphp/rector-src/pull/1471/commits/5a22f79351de192dfbec8e952f6fb986ad10469f

**Comment by samsonasik** [comment]
Now, the error is back to: 

```bash
There was 1 error:

1) Rector\Tests\Php74\Rector\Function_\ReservedFnFunctionRector\ReservedFnFunctionRectorTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
PHPStan\Parser\ParserErrorsException: Syntax error, unexpected T_FN, expecting T_STRING or '(' on line 9, Syntax error, unexpected '{', expecting T_DOUBLE_ARROW on line 10, Syntax error, unexpected T_LNUMBER, expecting T_VARIABLE on line 14
```

Do we need to skip test?

**Comment by TomasVotruba** [comment]
> Do we need to skip test?

Yes, for now so we can separate these 2 issues.

**Comment by samsonasik** [comment]
Ok, test now skipped.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**